### PR TITLE
#95/shuttle data

### PR DESCRIPTION
**Pull request by CarenRizk**
Created the Shuttle model, JSON file for shuttle data, and UI to display the list of shuttles.

### CODE REVIEW

File: app/src/main/java/com/example/concordia_campus_guide/Activities/MainActivity.java [hunk 1]
```diff
@@ -53,6 +57,10 @@ protected void onCreate(Bundle savedInstanceState) {
         swipeableInfoCard = BottomSheetBehavior.from(infoCard);
 
         locationFragment = (LocationFragment) getSupportFragmentManager().findFragmentById(R.id.locationFragment);
+
+        AppDatabase appDB = AppDatabase.getInstance(this);
+        List<Shuttle> jas = appDB.shuttleDao().getScheduleByCampusAndDay("SGW", "Friday");
```
Comment (Hambrsoom) [inline app/src/main/java/com/example/concordia_campus_guide/Activities/MainActivity.java]: What is the purpose of this here ? Was it only for testing? if yes, can you delete it ? and if no, I don't think this a  good place for it. Also, we should use proper variable names.

File: app/src/main/java/com/example/concordia_campus_guide/Activities/MainActivity.java [hunk 2]
```diff
@@ -53,6 +57,10 @@ protected void onCreate(Bundle savedInstanceState) {
...
+        String gs = "Asda" ;
```
Comment (Hambrsoom) [inline app/src/main/java/com/example/concordia_campus_guide/Activities/MainActivity.java]: I don't think this line is doing anything here. right? If yes, can you remove it?

File: app/src/main/java/com/example/concordia_campus_guide/Activities/MainActivity.java [hunk 3]
```diff
@@ -139,6 +147,12 @@ private void setUpDb(){
             Rooms rooms = ApplicationState.getInstance(this).getRooms();
             appDb.roomDao().insertAll(rooms.getRooms());
 
+            //Get shuttle schedule
+            Shuttles shuttles = ApplicationState.getInstance(this).getShuttles();
+            appDb.shuttleDao().insertAll(shuttles.getShuttles());
+
+            List<Shuttle> sh = appDb.shuttleDao().getScheduleByCampus("SGW");
```
Comment (Hambrsoom) [inline app/src/main/java/com/example/concordia_campus_guide/Activities/MainActivity.java]: Why are you retrieving the data for the shuttles in the SetUpDb() method? Is there a particular reason ? Because the  purpose of the setUpDb() method is just for inserting data inside the db.

File: app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivityViewModel.java [hunk 4]
```diff
@@ -39,4 +49,44 @@ public void setFrom(Place from) {
...
+            } else if (to.getClass() == Building.class) {
+                campusTo = ((Building) to).getCampus();
+            } else if (from.getClass() == Floor.class) {
+                campusFrom = ((Floor) from).getCampus();
+            } else if(to.getClass() == Floor.class) {
+                campusTo = ((Floor) to).getCampus();
+            }
+            if (campusFrom.compareTo(campusTo) == 0) {
+                return null;
+            }
```
Comment (Hambrsoom) [inline app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivityViewModel.java]: Can you elaborate a bit more about this if else statements and what their purpose is ?

File: app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivityViewModel.java [hunk 5]
```diff
@@ -39,4 +49,44 @@ public void setFrom(Place from) {
...
+        }
+
+        Calendar cal = Calendar.getInstance();
```
Comment (Hambrsoom) [inline app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivityViewModel.java]: Can you use calendar as variable name instead of cal? it would be clearer for the reader :)

File: app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivityViewModel.java [hunk 6]
```diff
@@ -39,4 +49,44 @@ public void setFrom(Place from) {
...
+
+    public String getShuttleDisplayText(List<Shuttle> shuttles) {
+        String content = "";
+        String campusTo = shuttles.get(0).getCampus().compareTo("SGW") == 0 ? "LOY" : "SGW";
+        if (shuttles == null || shuttles.size() == 0) {
+            return "No available routes using the shuttle service";
+        }
+        for (Shuttle shuttle : shuttles) {
+            content += shuttle.getCampus() + "  >   " + campusTo + ", leaves at: " + shuttle.getTime() + "\n";
+        }
```
Comment (Hambrsoom) [inline app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivityViewModel.java]: I would put this a separate method and call it later in here. What do you think ? It would make it easier to test these two methods later.

File: app/src/main/java/com/example/concordia_campus_guide/Models/RoomModel.java [hunk 7]
```diff
@@ -16,9 +16,7 @@
                         childColumns = {"floor_code"},
                         onDelete = CASCADE
                 ),
-                },
-        indices = {@Index(value = {"room_code"},
-        unique = true)})
+                })
```
Comment (Hambrsoom) [inline app/src/main/java/com/example/concordia_campus_guide/Models/RoomModel.java]: Why did you remove the room_code as being unique ?? Were you having some issues ?

File: app/src/main/res/raw/buildingcoordinates.json [hunk 8]
```diff
@@ -15,21 +15,24 @@
         "type": "Polygon",
         "coordinates": [
```
Comment (Hambrsoom) [inline app/src/main/res/raw/buildingcoordinates.json]: Buildingcoordinates is not used anymore. It will be deleted soon

File: app/src/main/java/com/example/concordia_campus_guide/Activities/MainActivity.java (hunk 1 above)
Comment (CarenRizk) [reply app/src/main/java/com/example/concordia_campus_guide/Activities/MainActivity.java]: Yeah, it was only for testing. I'll remove it. Good catch!

File: app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivityViewModel.java (hunk 4 above)
Comment (CarenRizk) [reply app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivityViewModel.java]: Because the from and to are places, so I can't get the campus since a place doesn't have campus. So I am checking if it is an instance of Floor or Building, then I am casting accordingly to be able to get the campus.

File: app/src/main/java/com/example/concordia_campus_guide/Models/RoomModel.java (hunk 7 above)
Comment (CarenRizk) [reply app/src/main/java/com/example/concordia_campus_guide/Models/RoomModel.java]: You were with me when we did this :P. Yes because "elevators", "womens" and "mens" are not unique.

File: app/src/test/java/com/example/concordia_campus_guide/LocationFragmentViewModelTest.java [hunk 9]
```diff
@@ -4,6 +4,7 @@
 
 import com.example.concordia_campus_guide.Fragments.LocationFragment.LocationFragmentViewModel;
 import com.example.concordia_campus_guide.Models.Building;
+import com.example.concordia_campus_guide.Models.Shuttle;
```
Comment (tarekait1996) [inline app/src/test/java/com/example/concordia_campus_guide/LocationFragmentViewModelTest.java]: why did we add this import here?

File: app/src/main/res/layout/routes_activity.xml [hunk 10]
```diff
@@ -70,15 +70,15 @@
...
                 android:layout_width="303dp"
                 android:layout_height="39dp"
                 android:layout_marginTop="8dp"
                 android:background="#29FFFFFF"
+                android:clickable="true"
                 android:ems="10"
                 android:inputType="textPersonName"
+                android:onClick="onClickFrom"
                 android:padding="8dp"
                 android:text="Name"
```
Comment (tarekait1996) [inline app/src/main/res/layout/routes_activity.xml]: Text should be written in strings.xml and referenced here, it will be easier that way when we support french text, we will just need to add the text in the french strings file with the same id.

File: app/src/main/java/com/example/concordia_campus_guide/Models/Floor.java [hunk 11]
```diff
@@ -70,5 +73,9 @@ public void setAltitude(float altitude) {
     public String getDisplayName(){
         return floorCode;
     }
+
+    public String getCampus() { return campus; }
```
Comment (tarekait1996) [inline app/src/main/java/com/example/concordia_campus_guide/Models/Floor.java]: This does not have to be in this PR specifically but please when updating/creating new files do not forget to update/create unit tests for the file. In this case, a floor unit test exists so maybe we should update it to test those methods?

File: app/src/main/java/com/example/concordia_campus_guide/Models/Campus.java [hunk 12]
```diff
@@ -0,0 +1,5 @@
+package com.example.concordia_campus_guide.Models;
+
+public enum Campus {
```
Comment (tarekait1996) [inline app/src/main/java/com/example/concordia_campus_guide/Models/Campus.java]: maybe we could use typedef, what do you think? by introducing more and more enums, it might slow our app. Look at this: https://android.jlelse.eu/android-performance-avoid-using-enum-on-android-326be0794dc3

File: app/src/main/java/com/example/concordia_campus_guide/Models/Floor.java (hunk 11 above)
Comment (CarenRizk) [reply app/src/main/java/com/example/concordia_campus_guide/Models/Floor.java]: Ok I updated the tests

File: app/src/main/java/com/example/concordia_campus_guide/Models/Campus.java (hunk 12 above)
Comment (CarenRizk) [reply app/src/main/java/com/example/concordia_campus_guide/Models/Campus.java]: Oh I'll delete this enum. I added it in the beginning and then decided to not use it. So now it's not being used anywhere

File: app/src/main/res/raw/buildingcoordinates.json (hunk 8 above)
Comment (CarenRizk) [reply app/src/main/res/raw/buildingcoordinates.json]: ok..

File: app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivityViewModel.java (hunk 6 above)
Comment (CarenRizk) [reply app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivityViewModel.java]: but the method is simple enough I don't think it needs to be more seperated

File: app/src/main/res/layout/routes_activity.xml (hunk 10 above)
Comment (CarenRizk) [reply app/src/main/res/layout/routes_activity.xml]: image

Comment (CarenRizk) [reply app/src/main/res/layout/routes_activity.xml]: This is actually not part of what I did I'm not sure how the PR sees it as a new line.. But whoever worked on it is actually retrieving it by ID and replacing name so I think it should be fine

File: app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivityViewModel.java (hunk 4 above)
Comment (thomasgauvin) [reply app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivityViewModel.java]: what about if the place is a room @CarenRizk ?

Comment (thomasgauvin) [reply app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivityViewModel.java]: I see that floor and building both have a campus. we should also consider this for room. i think it makes sense to add campus to place

File: app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivityViewModel.java [hunk 13]
```diff
@@ -39,4 +49,44 @@ public void setFrom(Place from) {
...
+        shuttles = appDB.shuttleDao().getScheduleByCampusAndDayAndTime(campusFrom, day, time.format(calendar.getTime()));
+        return shuttles;
+    }
+
+    public String getShuttleDisplayText(List<Shuttle> shuttles) {
+        String content = "";
+        if (shuttles == null || shuttles.size() == 0) {
+            return "No available routes using the shuttle service";
+        }
+        String campusTo = shuttles.get(0).getCampus().compareTo("SGW") == 0 ? "LOY" : "SGW";
```
Comment (thomasgauvin) [inline app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivityViewModel.java]: is this necessary? isn't getCampus() already returning "LOY" or "SGW"?

File: app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivityViewModel.java [hunk 14]
```diff
@@ -39,4 +49,44 @@ public void setFrom(Place from) {
...
+        String day = new SimpleDateFormat("EEEE", Locale.ENGLISH).format(calendar.getTime());
+        SimpleDateFormat time = new SimpleDateFormat("HH:mm");
+        shuttles = appDB.shuttleDao().getScheduleByCampusAndDayAndTime(campusFrom, day, time.format(calendar.getTime()));
+        return shuttles;
+    }
+
+    public String getShuttleDisplayText(List<Shuttle> shuttles) {
+        String content = "";
+        if (shuttles == null || shuttles.size() == 0) {
+            return "No available routes using the shuttle service";
```
Comment (thomasgauvin) [inline app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivityViewModel.java]: might make sense to put this string in class constants

File: app/src/main/java/com/example/concordia_campus_guide/Database/Converters/StringLongConverter.java [hunk 15]
```diff
@@ -0,0 +1,21 @@
+package com.example.concordia_campus_guide.Database.Converters;
+
+import androidx.room.TypeConverter;
+
+public class StringLongConverter {
```
Comment (thomasgauvin) [inline app/src/main/java/com/example/concordia_campus_guide/Database/Converters/StringLongConverter.java]: is this converter for long types or time types? we should make it clear in the class name

File: app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivityViewModel.java (hunk 13 above)
Comment (CarenRizk) [reply app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivityViewModel.java]: Yes but getCampus() is for the origin campus, this line is returning the destination campus. It's just for the UI

File: app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivityViewModel.java (hunk 4 above)
Comment (Hambrsoom) [reply app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivityViewModel.java]: I agree with thomas of removing the campus attribute from all the subclasses and leave it only in the place class. However, we will need to modify the json file of the rooms accordingly to add the campus key and values in there for each room

File: app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivityViewModel.java (hunk 6 above)
Comment (thomasgauvin) [reply app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivityViewModel.java]: agree with Caren here

File: app/src/main/java/com/example/concordia_campus_guide/Models/RoomModel.java (hunk 7 above)
Comment (thomasgauvin) [reply app/src/main/java/com/example/concordia_campus_guide/Models/RoomModel.java]: this will cause issues when we link walking points, we should discuss this

File: app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivityViewModel.java (hunk 13 above)
Comment (thomasgauvin) [reply app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivityViewModel.java]: oh so it inverts it
I get it cool

File: app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivityViewModel.java (hunk 6 above)
Comment (Hambrsoom) [reply app/src/main/java/com/example/concordia_campus_guide/Activities/RoutesActivityViewModel.java]: You guys have a point. I am good with that too

File: app/src/main/java/com/example/concordia_campus_guide/Models/RoomModel.java (hunk 7 above)
Comment (Hambrsoom) [reply app/src/main/java/com/example/concordia_campus_guide/Models/RoomModel.java]: I agree with thomas cuz I noticed that in the json file we are storing elevators (as the roomCode). Which might be confusing when finding the paths using walkingpoints
